File: code/Server/WorkoutGenerator.py
#!/usr/bin/env python3

from calendar import calendar
import random
import Days
import ExerciseAlt
import dataFetcher



#*
# * defines the workout generator class.
# * 
# * @author group4
# 
class WorkoutGenerator:
    RESTDAY = "Rest Day"
    RESTDESCRIPTION = "Enjoy your Rest"
    RESTCAT = 0
    RESTEQUIP = [0]

    #    *
    #     * generates workouts based on the current users selected preferences
    #     * 
    #     * @param preferences the current users selected preferences.
    #     * @return workoutCalendar the generated calendar
    #     
    def generateWorkouts(preferences):

        exerciseMap = WorkoutGenerator.populateExercises(preferences)
        calendar = {}
        for currentDay in Days.Days:
            if currentDay in preferences.getSelectedDays():
                calendar[currentDay.name] = WorkoutGenerator.createBalancedWorkout(exerciseMap)
            else:
                calendar[currentDay.name] = WorkoutGenerator.addRestDay()
        return calendar

    def addRestDay():
        test = {}
        restDay = []
        restDay.append(ExerciseAlt.ExerciseAlt.defineDetails(WorkoutGenerator.RESTDAY, WorkoutGenerator.RESTDESCRIPTION, WorkoutGenerator.RESTCAT, WorkoutGenerator.RESTEQUIP, 0))
        test["exercises"] = restDay
        return test


    def createBalancedWorkout(exerciseMap):
        test = {}
        workout = []
        for currentGroup in exerciseMap.keys():
            range = len(exerciseMap[currentGroup])
            print(range)
            amountToAdd = 3
            while amountToAdd > 0:
                randomIndex = int((random.random() * range))
                workout.append(exerciseMap[currentGroup][randomIndex])
                amountToAdd -= 1
        test["exercises"] = workout
        return test

    def populateExercises(preferences):
        exerciseMap = {}
        for current in preferences.getSelectedMuscles():
            exerciseMap[current.value] = dataFetcher.dataFetcher.getExercises(current.value, preferences.getIntensity())
        return exerciseMap

File: code/src/main/java/workout_manager/model/WorkoutGenerator.java
package workout_manager.model;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * defines the workout generator class.
 * 
 * @author group4
 */
public class WorkoutGenerator {
    private WorkoutDataFetcher dataFetcher;
    private static final String RESTDAY = "Rest Day";
    private static final String RESTDESCRIPTION = "Enjoy your Rest";
    private static final int RESTCAT = 0;
    private static final int[] RESTEQUIP = new int[] { 0 };

    /**
     * creates a workout generator object.
     */
    public WorkoutGenerator(WorkoutDataFetcher dataFetcher) {
        this.dataFetcher = dataFetcher;
    }

    /**
     * generates workouts based on the current users selected preferences
     * 
     * @param preferences the current users selected preferences.
     * @return workoutCalendar the generated calendar
     */
    public WorkoutCalendar generateWorkouts(Preferences preferences) {

        Map<MuscleGroup, ArrayList<ExerciseAlt>> exerciseMap = this.populateExercises(preferences);
        WorkoutCalendar workoutCalendar = new WorkoutCalendar();

        for (Days currentDay : Days.values()) {
            if (preferences.getSelectedDays().contains(currentDay)) {
                workoutCalendar.addWorkout(currentDay, this.createBalancedWorkout(exerciseMap));
            } else {
                workoutCalendar.addWorkout(currentDay, this.addRestDay());
            }

        }
        return workoutCalendar;
    }

    private Workout addRestDay() {
        Workout restDay = new Workout();
        restDay.addExercise(new ExerciseAlt(RESTDAY, RESTDESCRIPTION, RESTCAT, RESTEQUIP));
        return restDay;

    }

    private Workout createBalancedWorkout(Map<MuscleGroup, ArrayList<ExerciseAlt>> exerciseMap) {
        Workout workout = new Workout();
        for (MuscleGroup currentGroup : exerciseMap.keySet()) {
            int range = exerciseMap.get(currentGroup).size();
            int amountToAdd = 3;
            while (amountToAdd > 0) {
                int randomIndex = (int) (Math.random() * range);
                workout.addExercise(exerciseMap.get(currentGroup).get(randomIndex));
                amountToAdd--;
            }
        }
        return workout;
    }

    private Map<MuscleGroup, ArrayList<ExerciseAlt>> populateExercises(Preferences preferences) {
        Map<MuscleGroup, ArrayList<ExerciseAlt>> exerciseMap = new HashMap<MuscleGroup, ArrayList<ExerciseAlt>>();
        for (MuscleGroup current : preferences.getSelectedMuscles()) {
            exerciseMap.put(current, this.dataFetcher.getExercises(current.getCode()));
        }
        return exerciseMap;
    }

}
